src/backends/ucsi_debugfs.rs

```diff
@@ -0,0 +1,307 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// SPDX-FileCopyrightText: © 2024 Google
+// Ported from libtypec (Rajaram Regupathy <rajaram.regupathy@gmail.com>)
+
+//! The UCSI backend
+
+use std::io::Cursor;
+
+use crate::pd::Message;
+use crate::pd::MessageRecipient;
+use crate::pd::MessageResponseType;
+use crate::pd::Pdo;
+use crate::ucsi::AlternateMode;
+use crate::ucsi::CableProperty;
+use crate::ucsi::Capability;
+use crate::ucsi::Command;
+use crate::ucsi::ConnectorCapability;
+use crate::ucsi::ConnectorStatus;
+use crate::ucsi::GetAlternateModesRecipient;
+use crate::ucsi::PdoSourceCapabilitiesType;
+use crate::ucsi::PdoType;
+use crate::BcdWrapper;
+use crate::BitReader;
+use crate::Error;
+use crate::FromBytes;
+use crate::OsBackend;
+use crate::Result;
+use crate::ToBytes;
+
+/// A mere convenience to check if a response is null.
+trait NullResponse {
+    fn is_null(&self) -> bool;
+}
+
+impl NullResponse for Vec<u8> {
+    /// A convenience method to identify a null response.
+    fn is_null(&self) -> bool {
+        self.iter().all(|byte| *byte == 0)
+    }
+}
+
+mod driver {
+    use std::fs::File;
+    use std::io::Read;
+    use std::io::Seek;
+    use std::io::Write;
+    use std::os::fd::AsFd;
+
+    use crate::Error;
+    use crate::Result;
+
+    pub struct Driver {
+        /// The file descriptor used to send commands.
+        command_fd: File,
+        /// The file descriptor used to receive responses.
+        response_fd: File,
+    }
+
+    impl Driver {
+        pub fn new() -> Result<Self> {
+            let command_fd = std::fs::OpenOptions::new()
+                .write(true)
+                .open("/sys/kernel/debug/usb/ucsi/USBC000:00/command")?;
+
+            let mut response_fd = std::fs::OpenOptions::new()
+                .read(true)
+                .open("/sys/kernel/debug/usb/ucsi/USBC000:00/response")?;
+
+            response_fd.seek(std::io::SeekFrom::Start(0))?;
+
+            Ok(Self {
+                command_fd,
+                response_fd,
+            })
+        }
+
+        pub fn submit_command(&mut self, command: &[u8]) -> Result<usize> {
+            Ok(self.command_fd.write(command)?)
+        }
+
+        pub fn wait_response(&mut self) -> Result<Vec<u8>> {
+            let poll_fd =
+                nix::poll::PollFd::new(self.response_fd.as_fd(), nix::poll::PollFlags::POLLIN);
+            let timeout = nix::poll::PollTimeout::from(10000u16);
```

Comment:
It would be better to use a constant for 10000u16. I think it's 10_000 milliseconds, so 10 seconds?